include/engine/engine.hpp

```diff
@@ -30,38 +32,162 @@ namespace osrm
 namespace engine
 {
 
-class Engine final
+class EngineInterface
+{
+  public:
+    virtual ~EngineInterface(){};
+    virtual Status Route(const api::RouteParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Table(const api::TableParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Nearest(const api::NearestParameters &parameters,
+                           util::json::Object &result) const = 0;
+    virtual Status Trip(const api::TripParameters &parameters,
+                        util::json::Object &result) const = 0;
+    virtual Status Match(const api::MatchParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Tile(const api::TileParameters &parameters, std::string &result) const = 0;
+};
```

Comment:
Why the split again? For osrm <-> Engine pimpl? Can't you fwd. declare Engine, the unique_ptr will happily store a fwd. declared type - then in the impl. include the header.